_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -12,46 +12,64 @@ FFI tool, [Diplomat](https://github.com/rust-diplomat/diplomat).
+Manish, and originally implemented by Shadaj, a past Google intern, and was
+motivated by the lack of good FFI tools that satified ICU4X's requirements. This
+post focuses on how it works, but details on the motivation can be found in the
+[design document](https://github.com/rust-diplomat/diplomat/blob/main/docs/design_doc.md).
 
-Diplomat is a unidirectional FFI tool that allows programs in other languages
+Diplomat is a unidirectional[^1] FFI tool that allows programs in other languages
 to call Rust libraries via automatically generated, idiomatic APIs. At its core
 lays the abstract representation of types and methods, which represents the
 lowest common denominator for what types and methods look like across most
 modern programming languages. Since the abstract representation is so general,
 Diplomat is theoretically capable of generating bindings to any language that
-Rust has FFI support for, namely JavaScript via WebAssembly, and languages that
-support C FFI. Currently, it has support for generating bindings to C, C++, C#,
-and JavaScript/TypeScript.  Diplomat also generates native-feeling types and
+can support binding to C or WASM, since it generates `extern "C"` functions for
+each method. Currently, it has support for generating bindings to C, C++, C#,
+and JavaScript/TypeScript. Diplomat also generates native-feeling types and
 methods in the target language that call into the raw WASM/C interface.
 
-The cost of creating bindings to all these languages is just one set of FFI
-bindings, which are exposed through a module marked with the
-`#[diplomat::bridge]` proc macro. So if you have some large Rust library like
-ICU4X, you can attach Diplomat bindings and get native feeling APIs in every
-supported language at no extra cost.
+[^1]: Unidirectional FFI is when one language can invoke functions in another
+language, but not the other way around.
 
-# Abstract Representation
+The cost of creating bindings to all these languages is just one set of FFI API
+declarations, which looks mostly like normal Rust code exposed through a module
+marked with the `#[diplomat::bridge]` proc macro (inspired by `#[cxx::bridge]`).
+So if you have some large Rust library like ICU4X, you can attach Diplomat
+bindings and get native feeling APIs in every supported language at no extra cost.
+
+## Abstract Representation
+
+Since Diplomat aims to be language agnostic, it requires an abstract representation
+that makes sense to many languages. For this reason, it's abstract representation
+consists of only primitives (integers, booleans, strings), fieldless enums,
+structs, and methods. By supporting these ideas that are representable to most
+programming languages, Diplomat is able to reason about the types and generate
+conversion code between the two over FFI when calling Rust methods.
+
+The problem with this model is that sometimes we don't want to convert a value
+at the boundary because it's only ever used to call back into Rust code. This
```

Comment:
I'd say, more generally: "We don't want to convert a value at the boundary because its internals are just that, _internal_, and shouldn't escape the Rust-side implementation details. Rather, we want the other side to be able to call methods on it like a regular Rust API."